src/libstd/thread/local.rs

```diff
@@ -364,54 +398,84 @@ pub mod fast {
     impl<T> Key<T> {
         pub const fn new() -> Key<T> {
             Key {
-                inner: UnsafeCell::new(None),
-                dtor_registered: Cell::new(false),
-                dtor_running: Cell::new(false)
+                inner: LazyKeyInner::new(),
+                dtor_state: Cell::new(DtorState::Unregistered),
             }
         }
 
-        pub unsafe fn get(&self) -> Option<&'static UnsafeCell<Option<T>>> {
-            if mem::needs_drop::<T>() && self.dtor_running.get() {
-                return None
+        #[inline]
+        pub unsafe fn get<F: FnOnce() -> T>(&self, init: F) -> Option<&'static T> {
+            match self.inner.get() {
+                Some(val) => Some(val),
+                None => {
+                    if mem::needs_drop::<T>() {
```

Comment:
Could this conditional check happen inside a function called `try_initialize` and result in an early return? I suspect that would have similar codegen than the current pattern has

Reply:
Let me know if I interpreted this correctly.